### PR TITLE
Relay filter on home feed again

### PR DESCRIPTION
**Pull request by mplorentz**
## Issues covered
#1291

## Description
This code was merged, then reverted, then partially unreverted, and now finally is being fully unreverted. Along with some bugfixes. Clear as mud?

The first commit here is just a `git revert` and the code in there was already approved, so if you don't want to review that part again you can exclude commit `[c19ccb6](https://github.com/planetary-social/nos/pull/1337/commits/c19ccb625f2b7e5c684cd7b342afb20d908180f2)` from the diff, or maybe it's helfpul to review it separately.

The reason we reverted this was because it had a couple bugs:
1. The profile view would randomly scroll to the top.
2. When filtering the home feed by a relay, if you scrolled down fast enough you would only see notes from your follows.

The first bug was easily fixed by adding a `@State private var lastRefreshDate` to the ProfileView and using it the fetch request. This prevented the fetch request from changing slightly every time the ProfileView was re-rendered, which the PagedNoteListView would detect and call collectionView.reloadData().

The second bug was much trickier. The problem was that our paging code kind of assumed the feed was starting with no events. As you scroll down past 20 notes each relay is asked to load the next 20 notes. Well if you grab the scroll bar and scroll down 500 events then the relays are going to get asked for really old events and they will never catch up.

My solution was to detect this in `PagedRelaySubscription` and then put it into "catch up" mode, opening an extra subscription for all events from wherever you started your massive scroll down to what you are currently looking at. So far I haven't had much luck modifying the paging code without introducing new bugs, so please give this a close look.

## How to test
Try browsing the home feed, feeds of various relays, and the profile view. Try grabbing the scroll bar on the Mac and dragging it towards the bottom and see if it eventually grabs all the events you expect.

## Screenshots/Video
|Before|After|
|----|-----|
|Simulator Screenshot - iPhone 15 - 2024-07-12 at 13 33 07|simulator_screenshot_855B29F1-C556-4A53-8739-E317BF68B840|

### CODE REVIEW

**Comment by mplorentz** [comment]
Looks like swiftlint updated and introduced new rules. I'll fix the linting errors tomorrow but I'm still marking this ready for review so I can hopefully get some feedback tomorrow morning.

**Review by joshuatbrown — COMMENTED** [review]
A few thoughts for now. Full review coming tomorrow.

File: Nos/Views/PagedNoteListView.swift [hunk 1]
```diff
@@ -25,6 +25,8 @@ struct PagedNoteListView<Header: View, EmptyPlaceholder: View>: UIViewRepresenta
     /// by the `databaseFilter`. 
     let relayFilter: Filter
     
+    let relay: Relay?
```
Comment (joshuatbrown) [inline Nos/Views/PagedNoteListView.swift:28]: A doc comment would be nice.

**Review by joshuatbrown — COMMENTED** [review]
I think I can still get it into a state where notes are missing and just don't load. I scrolled super fast (dragging the scroll indicator down) and it seems like some notes are missing here between June 3 and July 10. I'm following Matt, Linda, Rabble, Nos, and my personal account -- I can't imagine that none of us posted between those dates.

Simulator Screenshot - iPhone SE (3rd generation) - 2024-07-25 at 10 01 44

**Comment by joshuatbrown** [comment]
I can get the same issue to happen -- missing posts in the middle -- by scrolling the Profile screen super fast (dragging the scroll indicator), then going back up. I'm currently looking at Rabble's profile and don't see posts between March 26 and July 14, though there are plenty before and after.

File: Nos/Service/Relay/Filter.swift [hunk 2]
```diff
@@ -5,13 +5,13 @@ import Foundation
...
     /// List of author identifiers the Filter should be constrained to.
-    let authorKeys: [RawAuthorID]
+    var authorKeys: [RawAuthorID]
 
     /// List of event identifiers the Filter should be constrained to.
     let eventIDs: [RawEventID]
 
     /// List of Note kinds to filter
-    let kinds: [EventKind]
+    var kinds: [EventKind]
```
[the remark this answers is not in the document]
Comment (mplorentz) [reply Nos/Service/Relay/Filter.swift:14]: 🤷‍♂️

File: Nos/Controller/PagedNoteDataSource.swift [hunk 3]
```diff
@@ -231,7 +250,11 @@ class PagedNoteDataSource<Header: View, EmptyPlaceholder: View>: NSObject, UICol
     /// `loadMoreIfNeeded(for:)` won't be called which means we'll never ask for the next page. So we need the timer.
     private func startAggressivePaging() {
         if aggressivePagingTimer == nil {
-            aggressivePagingTimer = Timer.scheduledTimer(withTimeInterval: 0.25, repeats: true) { [weak self] timer in
+            // Fire manually once because the timer doesn't fire immediately
+            let displayedDate = self.displayedDate(for: largestLoadedRowIndex) 
+            Task { await self.pager?.loadMore(displayingContentAt: displayedDate) }
```
[the remark this answers is not in the document]
Comment (mplorentz) [reply Nos/Controller/PagedNoteDataSource.swift:255]: Oh ugh. I knew in my head there should be a better way to do this. I asked the AI and it failed me. Thank you, real human.

**Comment by mplorentz** [comment]
[Linda wants](https://planetary-app.slack.com/archives/C04PLR5QFQQ/p1721922162967659?thread_ts=1721921374.391729&cid=C04PLR5QFQQ) to merge this with the scrolling bug as a known issue, so I'm marking this ready for review again.

**Review by joshuatbrown — APPROVED** [review]
This looks great! The one issue I pointed out before isn't easy to reproduce; I'm not super concerned about it and I know we're willing to ship with it anyway. Nice work!

**Pull request merged**